### PR TITLE
WIP documents for memory interface and value domain

### PR DESCRIPTION
**Pull request by RalfJung**
This also resolves https://github.com/rust-lang/unsafe-code-guidelines/pull/40. I concede to the majority opinion that the 4 terms in question are "value", "place", "value expression" and "place expression".

EDIT: Oh, I should also mention: just like the other WIP document, this is basically my personal proposal / opinion; I was hoping for a review to increase clarity but I do not think we should strive for consensus on the opinions expressed in those documents.

### CODE REVIEW

File: reference/src/glossary.md [hunk 1]
```diff
@@ -73,10 +73,44 @@ niches. For example, the "all bits uninitialized" is an invalid bit-pattern for
...
+* and loading a [value](#value) of the same type from it (through the place-to-value coercion).
+
+
+#### Value
+
+A *value* (called "value of the expression" or "rvalue" in C and "prvalue" in C++) is what gets stored in a [place](#place), and also the result of computing a [*value expression*][place-value-expr].
+A value has a type.
+For example, a value of type `u8` is a mathematical integer in the range `0..256`.
+Values can be (according to their type) turned into a list of bytes, which is called a [representation](#representation) of the value.
+Values are ephemeral; they arise during the computation of an instruction but are only ever persisted in memory through their representation.
```
Comment (Lokathor) [inline reference/src/glossary.md]: Does "ephemeral" fit here? Is there anything to a value other than its byte representation which exists during a computation and then doesn't get persisted in memory?

File: wip/value-domain.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        idx: u64,
+        data: Box<Self>,
+    },
+    /// A "bag of raw bytes".
+    RawBag(Vec<Byte<Pointer>>),
+    /* ... */
+}
+```
+
+As Rust grows, we might expand this definition. That is okay; all previously defined representation relations are still well-defined when the domain grows, the newly added values will just not be valid for old types as one would expect.
```
Comment (Lokathor) [inline wip/value-domain.md]: So, is it _complete for current rust_ and just not future-proof? Or is it incomplete _even in current rust_ and it's just WIP and don't worry about it?

File: reference/src/glossary.md (hunk 1 above)
Comment (gnzlbg) [reply reference/src/glossary.md]: Pointer provenance maybe ? AFAICT that gets persisted in memory but it is not part of their representation, although we haven't clarified this.

File: wip/memory-interface.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+* The *expression/statement semantics* of Rust boils down to explaining which "memroy events" (calls to the memory interface) happen in which order. This part of the specification is *pure* in the sense that it has no "state": everything that needs to be remembered from one expression evaluation to the next is communicated through memory.
+* The Rust *memory model* explains which interactions with the memory are legal (the others are UB), and which values can be returned by reads.
+
+The interface shown below is also opinionated in several ways.
+It is not intended to be able to support *any imaginable* memory model, but rather start the process of reducing the design space of what we consider a "reasonable" memory model for Rust.
+For example, it explicitly acknowledges that pointers are not just integers and that uninitialized memory is special (both are true for C and C++ as well but you have to read the standard very careful, and consult [non-normative defect report responses](http://www.open-std.org/jtc1/sc22/wg14/www/docs/dr_260.htm), to see this).
```
Comment (gnzlbg) [inline wip/memory-interface.md]: I thought defect reports were normative, even if the conclusion of the report is "nothing has to be done".

File: wip/memory-interface.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+Another key property of the interface presented below is that it is *untyped*.
+This implies that in Rust, *operations are typed, but memory is not*---a key difference to C and C++ with their type-based strict aliasing rules.
+At the same time, the memory model provides a *side-effect free* way to turn pointers into "raw bytes", which is *not* [the direction C++ is moving towards](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2364.pdf), and we might have to revisit this choice later if it turns out to not be workable.
+
+## Pointers
+
+One key question a memory model has to answer is *what is a pointer*.
+It might seem like the answer is just "an integer of appropriate size", but [that is not the case][pointers-complicated].
+This becomes even more prominent with aliasing models such as [Stacked Borrows].
+So the interface will leave it up to the concrete instance to answer this question, and carry `Pointer` as an associated type.
```
Comment (gnzlbg) [inline wip/memory-interface.md]: These reads like it expect multiple different Pointer types. It might be helpful to maybe mention this here somehow.

File: wip/memory-interface.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+/// fail, that means the program caused UB.
+trait Memory {
+    /// The type of pointer values.
+    type Pointer: Copy + PtrToInt;
+
+    /// The size of pointer values.
+    const PTR_SIZE: u64;
+
+    /// Create a new allocation.
+    fn allocate(&mut self, size: u64, align: u64) -> Result<Self::Pointer, Error>;
```
Comment (gnzlbg) [inline wip/memory-interface.md]: Is there a reason this doesn't use `std::alloc::Layout` and prefers `size` and `align` ?

File: wip/memory-interface.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+    /// Remove an allocation.
+    fn deallocate(&mut self, ptr: Self::Pointer, size: u64, align: u64) -> Result<(), Error>;
+
+    /// Write some bytes to memory.
+    fn write(&mut self, ptr: Self::Pointer, bytes: Vec<Byte<Self::Pointer>>) -> Result<(), Error>;
```
Comment (gnzlbg) [inline wip/memory-interface.md]: Why does bytes need to own the bytes here ?

File: wip/value-domain.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+The Rust value domain is described by the following (incomplete) type definition (in terms of an arbitrary `Pointer` type instantiated by [the memory interface][memory-interface]):
+
+```rust
+enum Value<Pointer> {
+    /// A mathematical integer.
+    Int(BigInt),
+    /// A Boolean value.
+    Bool(bool),
+    /// A pointer.
+    Ptr(Pointer),
```
Comment (gnzlbg) [inline wip/value-domain.md]: Are references pointers here ?

File: wip/value-domain.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    Uninit,
+    /// An n-tuple.
+    Tuple(Vec<Self>),
+    /// A variant of a sum type.
+    Variant {
+        idx: u64,
+        data: Box<Self>,
+    },
+    /// A "bag of raw bytes".
+    RawBag(Vec<Byte<Pointer>>),
```
Comment (gnzlbg) [inline wip/value-domain.md]: Why is this needed ? An array should be able to cover this.

File: wip/value-domain.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    /// A mathematical integer.
+    Int(BigInt),
+    /// A Boolean value.
+    Bool(bool),
+    /// A pointer.
+    Ptr(Pointer),
+    /// An uninitialized value.
+    Uninit,
+    /// An n-tuple.
+    Tuple(Vec<Self>),
```
Comment (gnzlbg) [inline wip/value-domain.md]: Does this covers tuple, tuple structs, structs, and arrays ?

File: reference/src/glossary.md [hunk 10]
```diff
@@ -73,10 +73,44 @@ niches. For example, the "all bits uninitialized" is an invalid bit-pattern for
...
 
 ### TODO
 
 * *tag*
-* *rvalue*
-* *lvalue*
-* *representation*
+
+[value-domain]: https://github.com/rust-lang/unsafe-code-guidelines/tree/master/wip/value-domain.md
+[place-value-expr]: https://doc.rust-lang.org/reference/expressions.html#place-expressions-and-value-expressions
```
Comment (gnzlbg) [inline reference/src/glossary.md]: I think these should be repeated here, for completeness.

File: wip/value-domain.md [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+## Values
+
+The Rust value domain is described by the following (incomplete) type definition (in terms of an arbitrary `Pointer` type instantiated by [the memory interface][memory-interface]):
+
+```rust
+enum Value<Pointer> {
+    /// A mathematical integer.
+    Int(BigInt),
+    /// A Boolean value.
+    Bool(bool),
```
Comment (Centril) [inline wip/value-domain.md]: Ostensibly it would be more precise to define `Int` and `Bool` using a sigma type with `({x: Nat | x ∈ start..end}, start: Nat, end: Nat)` or using [`Fin : (n : Nat) -> Type`](https://github.com/idris-lang/Idris-dev/blob/master/libs/base/Data/Fin.idr#L11) but well... can't use Rust for notation anymore. =P

File: wip/value-domain.md (hunk 7 above)
Comment (Centril) [inline wip/value-domain.md]: Also what about our friend `#[rustc_layout_scalar_valid_range_start(N)]`?

File: wip/value-domain.md [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    /// A pointer.
+    Ptr(Pointer),
+    /// An uninitialized value.
+    Uninit,
+    /// An n-tuple.
+    Tuple(Vec<Self>),
+    /// A variant of a sum type.
+    Variant {
+        idx: u64,
```
Comment (Centril) [inline wip/value-domain.md]: Shouldn't this be abstracted away? `discriminant_value` returns this today but that might not necessarily hold tomorrow.

File: wip/value-domain.md [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+A value `Tuple([t, u])` is represented by a list of bytes `rt ++ pad1 ++ ru ++ pad2` (using `++` for list concatenation) if:
+
+* `t` is represented by `rt` at type `T`.
+* `u` is represented by `ru` at type `U`.
+* The length of `rt ++ pad1` is equal to the offset of the `U` field.
+* The length of the entire list `rt ++ pad1 ++ ru ++ pad2` is equal to the size of `Pair<T, U>`.
+
+This relation specifies that values of type `Pair` are always 2-tuples (aka, pairs).
+It also says that the actual content of the padding bytes is entirely irrelevant, we only care to have the right number of them to "pad" `ru` to the right place and to "pad" the entire list to have the right length.
+So, for example when considering `Pair<u8, u16>`, the value `Tuple[42, 119]` is represented on a little-endian target by `[Raw(42), byte, Raw(119), Raw(0)]` for *any* `byte: Byte`.
```
Comment (Centril) [inline wip/value-domain.md]: Speaking of... is endianess an associated type of "`Machine`"?

File: wip/value-domain.md (hunk 2 above)
Comment (RalfJung) [reply wip/value-domain.md]: It's not complete for anything.

**Comment by RalfJung** [comment]
Wow, so much feedback so quick! Thanks a lot. I hope I can go over it tonight.

One thing I should have mentioned: the part in the glossary is hopefully uncontroversial, we should agree on it for merging. The "WIP" part is, just like the other WIP document, basically my personal proposal / opinion; I was hoping for a review to increase clarity but I do not think we should strive for consensus on the opinions expressed in those documents.

File: reference/src/glossary.md (hunk 1 above)
Comment (RalfJung) [reply reference/src/glossary.md]: The provenance of a pointer is part of `Pointer`.

File: wip/memory-interface.md (hunk 3 above)
Comment (RalfJung) [reply wip/memory-interface.md]: From what I was told, they are normative for the standard they were reported against, but when reading C18 I don't have to read all defect reports of all prior versions. That would be ludicrous.

C18 contains the exact same wording as C11 for indeterminate values.

I can still remove the "non-normative" part of the sentence, it doesn't really add anything.

File: wip/memory-interface.md (hunk 5 above)
Comment (RalfJung) [reply wip/memory-interface.md]: Fewer dependencies.

File: wip/memory-interface.md (hunk 6 above)
Comment (RalfJung) [reply wip/memory-interface.md]: This is more pseudo-code than code. Using a different type for the argument of `read` and the result of `write` seemed silly.

File: wip/value-domain.md (hunk 7 above)
Comment (RalfJung) [reply wip/value-domain.md]: Yes, this is about pointer *values* which are the values of both the reference and raw pointer *types*. I will clarify.

File: wip/value-domain.md (hunk 8 above)
Comment (RalfJung) [reply wip/value-domain.md]: I don't follow. No other variant of `Value` even lets you have a `Byte`.

File: reference/src/glossary.md (hunk 10 above)
Comment (RalfJung) [reply reference/src/glossary.md]: What should be repeated?

File: reference/src/glossary.md [hunk 14]
```diff
@@ -73,10 +73,44 @@ niches. For example, the "all bits uninitialized" is an invalid bit-pattern for
 `&mut T`, but this bit-pattern cannot be used by layout optimizations, and is not a
 niche.
 
+#### Place
+
+A *place* (called "lvalue" in C and "glvalue" in C++) is the result of computing a [*place expression*][place-value-expr].
+A place definitely stores some location in memory (a "pointer"), and might contain more information such as size or alignment (the details will have to be determined as the Rust Abstract Machine gets specified more precisely).
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply reference/src/glossary.md]: I don't understand the distinction.

File: wip/memory-interface.md [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+
+```rust
+enum Byte<Pointer> {
+    /// The "normal" case: a (frozen, initialized) integer in `0..256`.
+    Raw(u8),
+    /// An uninitialized byte.
+    Uninit,
+    /// One byte of a pointer.
+    PtrFragment {
+        /// The pointer of which this is a byte.
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply wip/memory-interface.md]: Suggestions? I cannot come up with another phrasing that seems any less complex.

File: wip/memory-interface.md [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+}
+```
+
+The purpose of `PtrFragment` is to be able to have a byte-wise representation of a `Pointer`.
+On a 32-bit system, the sequence of 4 bytes representing `ptr: Pointer` is:
+```
+[PtrFragment { ptr, idx: 0 }, PtrFragment { ptr, idx: 1 }, PtrFragment { ptr, idx: 2 }, PtrFragment { ptr, idx: 3 }]
+```
+
+Based on the `PtrToInt` trait (see below), we can turn every initialized `Byte` into an integer in `0..256`:
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply wip/memory-interface.md]: That's why I said *initialized* `Byte`.

File: wip/memory-interface.md [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+/// actually do change the current state when considering concurrency or
+/// Stacked Borrows.
+/// And finally, all operations are fallible (they return `Result`); if they
+/// fail, that means the program caused UB.
+trait Memory {
+    /// The type of pointer values.
+    type Pointer: Copy + PtrToInt;
+
+    /// The size of pointer values.
+    const PTR_SIZE: u64;
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply wip/memory-interface.md]: Maybe? I think putting it in memory for now works just as well.

File: reference/src/glossary.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (Centril) [reply reference/src/glossary.md]: It seems to me "memory location" refers to the cells holding the data itself and "stores some location" is a pointer to those cells? -- a question of indirection that is?

File: wip/memory-interface.md [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+    }
+}
+```
+
+## Memory interface
+
+The Rust memory interface is described by the following (not-yet-complete) trait definition:
+
+```rust
+/// *Note*: All memory operations can be non-deterministic, which means that
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply wip/memory-interface.md]: This is about the semantics of run-time Rust. Why are you asking about `const fn`?

File: wip/memory-interface.md [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+/// executing the same operation on the same memory can have different results.
+/// We also let all operations potentially mutate memory. For example, reads
+/// actually do change the current state when considering concurrency or
+/// Stacked Borrows.
+/// And finally, all operations are fallible (they return `Result`); if they
+/// fail, that means the program caused UB.
+trait Memory {
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply wip/memory-interface.md]: Agda is way too much like Haskell to be readable. ;)

File: wip/memory-interface.md [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+The Rust memory interface is described by the following (not-yet-complete) trait definition:
+
+```rust
+/// *Note*: All memory operations can be non-deterministic, which means that
+/// executing the same operation on the same memory can have different results.
+/// We also let all operations potentially mutate memory. For example, reads
+/// actually do change the current state when considering concurrency or
+/// Stacked Borrows.
+/// And finally, all operations are fallible (they return `Result`); if they
+/// fail, that means the program caused UB.
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply wip/memory-interface.md]: That just adds noise though.  So no I think that's not a good idea.

File: wip/value-domain.md (hunk 11 above)
Comment (RalfJung) [reply wip/value-domain.md]: I also don't think I agree. If this was Coq I'd use `Z`, not some `Fin` type, for `Int`.

File: wip/memory-interface.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (Centril) [reply wip/memory-interface.md]: Ah woops; read too fast!

File: wip/value-domain.md (hunk 13 above)
Comment (RalfJung) [reply wip/value-domain.md]: Maybe.

File: reference/src/glossary.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (RalfJung) [reply reference/src/glossary.md]: In my terminology, "location" is what we call the value of ML's `ref` types.

It *identifies* a cell that holds data, but it is not itself that cell.
A `Pointer` is basically location + provenance.

File: wip/memory-interface.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (Centril) [reply wip/memory-interface.md]: Well a `const fn` could execute at run-time as well and we could ostensibly tell LLVM about the deterministic aspects of it?

[the remark this answers is not in the document]
Comment (RalfJung) [reply wip/memory-interface.md]: Maybe, but what does that comment have to do with a basic doc comment about the *interface* of run-time memory?

One day when we have not just an interface but also an implementation of it (i.e., a memory model), and we are writing a document where we formally argue why our compilation to LLVM is correct... *then* I think your question becomes relevant. But I don't see how its answer affects the document we are discussing here.

File: reference/src/glossary.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (Centril) [reply reference/src/glossary.md]: Ah; can we clarify this distinction (I don't have suggestions... :D)?

File: wip/memory-interface.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (Centril) [reply wip/memory-interface.md]: Maybe just add:
```suggestion
        /// The pointer of which this is a byte.
        /// That is, the byte is a fragment of this pointer.
```
?

File: wip/memory-interface.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (Centril) [reply wip/memory-interface.md]: Alternatively, maybe write `-> Result<Stuff, UndefinedBehavior>`?

File: wip/memory-interface.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (Centril) [reply wip/memory-interface.md]: > Maybe, but what does that comment have to do with a basic doc comment about the _interface_ of run-time memory?

Fair enough; With emphasis on *"can"* I suppose that within a `const fn` context any non-deterministic an implementation would return `Error` then.

File: wip/value-domain.md (hunk 11 above)
Comment (Centril) [reply wip/value-domain.md]: Ah I see. That makes sense, the range would be more appropriate for the value relation for a type instead. Tho... in that case, why do we need `Bool` at all and why is that not subsumed by `Int`?

File: wip/memory-interface.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (RalfJung) [reply wip/memory-interface.md]: That, or the non-determinism must not be observable, or something like that.

The details of what actually *is* non-deterministic have to be filled in my the memory model anyway.

File: wip/value-domain.md (hunk 11 above)
Comment (RalfJung) [reply wip/value-domain.md]: Are you a set theorist or what? ;) "boolean" and "integer" are two distinct mathematical concepts.

Comment (Centril) [reply wip/value-domain.md]: Sure, but why is this distinction made here rather than when relating values to `bool`?

File: wip/memory-interface.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (Centril) [reply wip/memory-interface.md]: My thinking was that there are going to be more configurations like this so accumulating then in one place could be useful?

File: wip/value-domain.md (hunk 11 above)
Comment (RalfJung) [reply wip/value-domain.md]: Because the entire point of `Value` is to capture the mathematical concepts that are represented by the data we store in memory. If there is no concept of a "boolean" in `Value`, then the language literally does not have the concept of a "boolean" anywhere.

File: reference/src/glossary.md (hunk 1 above)
Comment (gnzlbg) [reply reference/src/glossary.md]: I think this is made clearer below, where it is shown that bytes can belong to a pointer, and therefore provenance is part of the representation.

File: wip/memory-interface.md (hunk 3 above)
Comment (gnzlbg) [reply wip/memory-interface.md]: > From what I was told, they are normative for the standard they were reported against, but when reading C18 I don't have to read all defect reports of all prior versions. That would be ludicrous.

Sounds correct to me. The changes proposed in defect reports are automatically merged into the standard draft, so they become part of the next standard version automatically. That particular report is a bit weird, in that its conclusion is that the defect isn't a defect because the standard is clear and therefore nothing must be done. If people have to fill a defect report, and reference it everywhere, the standard is obviously not clear..

File: reference/src/glossary.md (hunk 10 above)
Comment (gnzlbg) [reply reference/src/glossary.md]: I meant the place-expression and value-expression part of the reference.

File: wip/value-domain.md (hunk 8 above)
Comment (gnzlbg) [reply wip/value-domain.md]: Oh, I thought that BigInt would let you have a byte. So if aI cast a pointer to an `usize`, is that usize a `BigInt` or a `RawBag` ?

File: reference/src/glossary.md (hunk 10 above)
Comment (RalfJung) [reply reference/src/glossary.md]: I see. Seems weird to me to duplicate parts of the reference?

File: reference/src/glossary.md [hunk 21]
```diff
@@ -73,10 +73,44 @@ niches. For example, the "all bits uninitialized" is an invalid bit-pattern for
...
+A place has a type.
+Places are ephemeral; they arise during the computation of an instruction but cannot be persisted in memory.
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply reference/src/glossary.md]: Not sure which one "above" you mean (depending on the page these are displayed in a different order), but... statics are a good point. I had not thought about them.

[the remark this answers is not in the document]
Comment (Centril) [reply reference/src/glossary.md]: Argh... by "above" I mean "below" =P

(https://github.com/rust-lang/unsafe-code-guidelines/pull/175#discussion_r307252771)

File: reference/src/glossary.md (hunk 1 above)
Comment (RalfJung) [reply reference/src/glossary.md]: The difference between a value and its representation is comparable to the difference between run-time data and serialized data. The run-time data is also ephemeral in some sense; only serialized data is persisted across machine reboots, for example.

File: reference/src/glossary.md [hunk 22]
```diff
@@ -73,10 +73,44 @@ niches. For example, the "all bits uninitialized" is an invalid bit-pattern for
...
+
+#### Value
+
+A *value* (called "value of the expression" or "rvalue" in C and "prvalue" in C++) is what gets stored in a [place](#place), and also the result of computing a [*value expression*][place-value-expr].
+A value has a type.
+For example, a value of type `u8` is a mathematical integer in the range `0..256`.
+Values can be (according to their type) turned into a list of bytes, which is called a [representation](#representation) of the value.
+Values are ephemeral; they arise during the computation of an instruction but are only ever persisted in memory through their representation.
+
+#### Representation (relation)
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply reference/src/glossary.md]: I was afraid going too formal would drive away too many readers.

File: wip/memory-interface.md (hunk 3 above)
Comment (RalfJung) [reply wip/memory-interface.md]: Agreed, I think the C committee's position here is harmful. But alas, that is not something we are going to fix here and now.

File: wip/memory-interface.md (hunk 4 above)
Comment (RalfJung) [reply wip/memory-interface.md]: @gnzlbg  I do not understand what you mean. This works like any other associated type: the memory model defines which type this is *actually* going to be.

File: wip/memory-interface.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (RalfJung) [reply wip/memory-interface.md]: Yeah I was going to do this once we have more than a single item.

File: wip/value-domain.md (hunk 8 above)
Comment (RalfJung) [reply wip/value-domain.md]: I meant `Byte` as in what the memory-interface defines. `BigInt` is a mathematical integer, as the comment indicates. So when you cast a ptr to an int, you get a mathematical integer and that is represented via the `Int` variant.

`RawBag` is for unions.

File: wip/memory-interface.md (hunk 4 above)
Comment (gnzlbg) [reply wip/memory-interface.md]: If it were an associated type, it would be a `type Pointer = ...;` field. But making it an input type parameter gives the impression that one can instantiate this with different Pointer types, for the same memory model.

Comment (RalfJung) [reply wip/memory-interface.md]: Oh. The intention is to instantiate this with `Memory::Pointer` but we cannot do that yet because we need to define `Byte` to even define the trait.

I will try to clarify.

File: wip/value-domain.md (hunk 7 above)
Comment (RalfJung) [reply wip/value-domain.md]: Good question.

I don't think that will change anything about `Value`, but it is worth giving an example representation relation for it.

File: wip/value-domain.md (hunk 8 above)
Comment (RalfJung) [reply wip/value-domain.md]: I added a note saying `RawBag` is for unions, and hope this answers your question.

**Comment by RalfJung** [comment]
Okay, I think I responded to all comments and edited the document accordingly. Thanks a ton for all your feedback!

File: wip/memory-interface.md [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+[provenance]: https://github.com/rust-lang/unsafe-code-guidelines/blob/master/reference/src/glossary.md#pointer-provenance
+
+## Bytes
+
+The unit of communication between the memory model and the rest of the program is a *byte*.
+Again, the question of "what is a byte" is not as trivial as it might seem; beyond `u8` values we have to represent `Pointer`s and [uninitialized memory][uninit].
+We define the `Byte` type as follows, where `Pointer` will later be instantiated with the `Memory::Pointer` associated type.
+
+```rust
+enum Byte<Pointer> {
```
Comment (RalfJung) [inline wip/memory-interface.md]: Maybe `Byte` is not a good name for this, as it can be viewed as re-defining the term. I am basically starting from "a byte is what gets stored at a memory address", but other people might think "a byte is 8 bits" or "a byte is an integer in `0..256`". and then this type here makes no sense. It turns out that matching all these definitions at once is just not possible, though.

**Comment by RalfJung** [comment]
It's been two weeks, and this is just a WIP document. Any objections to merging?

**Comment by Centril** [comment]
Let's do it!
